_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,375 @@
+Jupiter's context with `MockitoExtension.class`.
+
+This makes your setup even neater and conciser!
+
+### Mockito Spring Style
+
+```java
+
+@ExtendWith(SpringExtension.class)
+class CityServiceImplTestMockitoSpringStyle {
+
+    // System under Test (SuT)
+    private CityService cityService;
+
+    // Mocks
+    @MockBean
+    private CityRepository cityRepository;
+
+    @BeforeEach
+    void setUp() {
+        cityService = new CityServiceImpl(cityRepository);
+    }
+
+}
+```
+
+If you have a more complex test fixture, and you want to inject the mock into Spring's `ApplicationContext` you can make
+use of `@MockBean`. But caution: `@MockBean` is not an annotation from Mockito but from Spring!
+
+## Mockito in Wilderness
+
+```java
+class CityServiceImplTestMockitoDonts {
+
+    // System under Test (SuT)
+    private CityService cityService;
+
+    // Mocks
+    private CityRepository cityRepository;
+
+    // Helper
+    private City expected;
+
+    @BeforeEach
+    void setUp() {
+        expected = createCity();
+        cityRepository = Mockito.mock(CityRepository.class);
+        cityService = new CityServiceImpl(cityRepository);
+
+        // Avoid such complex declarations
+        Mockito.when(cityRepository.save(expected))
+                .thenReturn(Optional.of(expected));
+        Mockito.when(cityRepository.find(expected.getId()))
+                .thenReturn(Optional.of(expected));
+        Mockito.when(cityRepository.findByName(expected.getName()))
+                .thenReturn(Optional.of(expected));
+        Mockito.when(cityRepository.findAllByCanton(expected.getCanton()))
+                .thenReturn(Collections.singleton(expected));
+        Mockito.when(cityRepository.findAllByCountry(expected.getCanton().getCountry()))
+                .thenReturn(Collections.singleton(expected));
+    }
+
+    @Test
+    void save() throws ElementNotFoundException {
+        ReflectionAssert.assertReflectionEquals(expected, cityService.save(expected));
+    }
+
+    @Test
+    void find() throws ElementNotFoundException {
+        ReflectionAssert.assertReflectionEquals(expected, cityService.find(expected.getId()));
+    }
+
+    @Test
+    void delete() throws ElementNotFoundException {
+        cityService.delete(expected);
+        Mockito.verify(cityRepository).delete(expected);
+    }
+
+    @Test
+    void findByName() throws ElementNotFoundException {
+        ReflectionAssert.assertReflectionEquals(expected, cityService.findByName(expected.getName()));
+    }
+
+}
+```
+
+Knowing how to create the mocks, let's have a look at what you should avoid. Otherwise, you test cases are confusing
+what decreases your code's maintainability. It will save you much time debugging and doesn't let your team members guess
+what the intent of the test case is.
+
+* Do not concentrate all the Mockito declarations in the `setUp` method! Even tough the test cases are reduced to a
```

Comment:
These 5 bullet points are gold for testing beginners! To make them stand out, I would like to see each of them in their own sub section with their own small, focused code example. Then you can remove the big code example above (which is a bit disconnected anyway, because it only applies to a few of the bullet points). Added benefit: readers don't have to scroll back and forth between the best practice and the code example.